pkg/operators/wasm/maps.go

```diff
@@ -0,0 +1,208 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"context"
+
+	"github.com/cilium/ebpf"
+	"github.com/tetratelabs/wazero"
+	wapi "github.com/tetratelabs/wazero/api"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+func (i *wasmOperatorInstance) addMapFuncs(env wazero.HostModuleBuilder) {
+	exportFunction(env, "getMap", i.getMap,
+		[]wapi.ValueType{wapi.ValueTypeI64}, // MapName
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Map
+	)
+
+	exportFunction(env, "mapLookup", i.mapLookup,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // Map
+			wapi.ValueTypeI64, // Key pointer
+			wapi.ValueTypeI64, // Value pointer
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "mapUpdate", i.mapUpdate,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // Map
+			wapi.ValueTypeI64, // Key pointer
+			wapi.ValueTypeI64, // Value pointer
+			wapi.ValueTypeI64, // Flag
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+
+	exportFunction(env, "mapDelete", i.mapDelete,
+		[]wapi.ValueType{
+			wapi.ValueTypeI32, // Map
+			wapi.ValueTypeI64, // Key pointer
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+}
+
+// getMap gets an existing map.
+// Params:
+// - stack[0] is the name of the map (string encoded)
+// Return value:
+// - Map handle on success, 0 on error
+func (i *wasmOperatorInstance) getMap(ctx context.Context, m wapi.Module, stack []uint64) {
+	mapNamePtr := stack[0]
+
+	mapName, err := stringFromStack(m, mapNamePtr)
+	if err != nil {
+		i.logger.Warnf("getMap: reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+
+	ebpfMap, ok := i.gadgetCtx.GetVar(operators.MapPrefix + mapName)
+	if !ok {
+		i.logger.Warnf("get map: no map for name %q", mapName)
+		stack[0] = 0
+		return
+	}
+
+	ebpfMap, ok = ebpfMap.(*ebpf.Map)
+	if !ok {
+		i.logger.Warnf("get map: map %q is not an ebpf map", mapName)
+		stack[0] = 0
+		return
+	}
+
+	stack[0] = wapi.EncodeU32(i.addHandle(ebpfMap))
+}
+
+// mapLookup searches the map for a value corresponding to the given key.
+// Params:
+// - stack[0]: Map handle
+// - stack[1]: Key pointer
+// - stack[2]: Value pointer
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) mapLookup(ctx context.Context, m wapi.Module, stack []uint64) {
+	mapHandle := wapi.DecodeU32(stack[0])
+	keyPtr := stack[1]
+	valuePtr := stack[2]
+
+	ebpfMap, ok := getHandle[*ebpf.Map](i, mapHandle)
```

Comment:
Thinking aloud about race conditions:

I am looking at `pkg/operators/ebpf/ebpf.go` `ebpfInstance.Close()`: it calls `i.collection.Close()`, which will call the `close()` syscall on the map fd.

And below, the call to `ebpfMap.LookupBytes()` cannot check if the map fd has been closed (or even worse, the fd number reallocated to another ebpf map).

Do we have any guarantee that `ebpfInstance.Close()` is only called after the wasm module is stopped?

cc @mauriciovasquezbernal : similar kind of topic as in #3542.

Reply:
> And below, the call to ebpfMap.LookupBytes() cannot check if the map fd has been closed (or even worse, the fd number reallocated to another ebpf map).

The cilium/ebpf library sets the FD of the map to -1 on Close() and returns `math.MaxUint32`, hence the kernel will return an error there. It's still an issue, but there is not a risk of using the wrong map.

> Do we have any guarantee that ebpfInstance.Close() is only called after the wasm module is stopped?

Unfortunately we don't have any. @burak-ok was suggesting the image layer operators should be stopped in reverse order, that will call wasm.Stop() before ebpf.Stop(). I think that's a good idea. However the order of the operators is [determined by the layers' order](https://github.com/inspektor-gadget/inspektor-gadget/blob/146c28a290bb1a1d14f1aff1acdcfed88681a80a/pkg/operators/oci-handler/oci.go#L310), so we'll need to sort those operators based on a priority to avoid it depending on the image.